### DISCUSSION TITLE
Navigation Problem

### DISCUSSION THREAD

**Discussion by anresa** [neuefische/questions#42] [Web] [answered]
Hallo,

ich will, dass meine Navigation animiert ist und der Kreis je nach Seite mitgeht. So:

<img width="377" alt="Bildschirmfoto 2022-08-02 um 15 37 20" src="https://user-images.githubusercontent.com/106099037/182393130-b459b115-8e26-428f-ba64-96dc2ca0fef3.png">

<img width="377" alt="Bildschirmfoto 2022-08-02 um 15 36 43" src="https://user-images.githubusercontent.com/106099037/182393143-92410e4c-e487-4280-972e-88b5d6fccfeb.png">

Zwischen drin hat das auch mal geklappt. Jetzt geht es nicht mehr und ich kann nicht rausfinden woran es liegt. Was hab ich bisher versucht? Alles neu gemacht, seit Stunden getestet, einzelne Komponenten ein- und ausgeblendet. Nichts funktioniert, die Animation reagiert nicht mehr.

Hier ist mein Pull Request mit dem ganzen Code dazu: [Navigation](https://github.com/anresa/capstone-project/pull/11)

Danke für eure Hilfe.

**Comment by ThomasF85** [comment, marked as answer]
Hallo,

ich kann bei dir ein kritisches Problem feststellen:
In deiner App.css hast du folgende Regel definiert:

```
* {
  ...
  width: 375px;
}
```

Das führt dazu dass jedes HTML Element die Breite 375px bekommt (außer es gibt eine weitere Regel die Breite definiert), unter anderem auch die svg Icons die du in der Navigation benutzt. Dadurch überlappen sich deine Links. Für den Nutzer sieht es so aus als würde er auf einen Link klicken, tatsächlich klickt er aber auf einen anderen Link. Das Problem erkennst du wenn du in den dev-tools des Browsers deine Elemente "inspectest". Entferne diese Regel bitte zuerst.

Dass die Animation nicht funktioniert liegt an der Herangehensweise. Deine Seite wird nach jedem Klick auf die Navigation komplett neu geladen, also werden auch immer wieder die folgenden Zeilen Code ausgeführt:
```
          <li className="list active">
              <a href="/add">
                <IoIosAddCircleOutline className="icon" />
                <span className="text">Add</span>
              </a>
           </li>
```
Dadurch ist immer wieder an der gleichen Stelle der Liste die CSS Klasse active gesetzt.
Du versuchst hier direkte DOM-Manipulationen vorzunehmen (.querySelectorAll(), .classList.add(), .classList.remove()). Bitte nimm keine solchen direkten DOM-Manipulationen in React vor. Versuche stattdessen mit State und Props zu arbeiten und definiere in deinen Komponenten wie sich diese Verhalten und wie diese aussehen sollen, je nach State und Props.
Eine gute Möglichkeit deine Navigation umzusetzen ist es anstatt a-Tags die Komponente NavLink von react-router-dom zu nutzen. Diese Komponente "weiß" ob sie gerade aktiv ist, wenn zB ein solcher NavLink auf die Seite "favorites" verlinkt, dann "weiß" der Link ob du gerade auf der Seite "favorites" bist oder nicht. Du musst dich also garnicht darum kümmern ob der Link aktiv ist oder nicht. Du kannst dem Links styles oder CSS Klassen mitgeben, die nur dann angewendet werden wenn der Link aktiv ist. Hier findest du die [docs](https://reactrouter.com/docs/en/v6/components/nav-link).

**Reply by anresa** [reply]
@ThomasF85 @pixelass  Vielen Dank, das hilft schon sehr! 
Ich habe den Code entsprechend umgebaut. Ich benötige noch Hilfe um die CSS Klassen einzubinden/ aktiv zu setzen. 

### Navigation.js
```
import './Navigation.css';
import {NavLink} from 'react-router-dom';
import {IoIosAddCircleOutline} from 'react-icons/io';
import {IoCopyOutline} from 'react-icons/io5';
import {IoIosMenu} from 'react-icons/io';
import {IoIosHeartEmpty} from 'react-icons/io';
import {IoPersonOutline} from 'react-icons/io5';

export default function Navigation() {
  let activeStyle = {
    textDecoration: 'underline',
  };


  return (
    <>
      <div className="navigation">
        <section className="navigation-box">
          <ul>
            <li>
              <NavLink to="add" style={({isActive}) => (isActive ? activeStyle : undefined)}>
                <IoIosAddCircleOutline className="icon" />
                <span className="text">Add</span>
              </NavLink>
            </li>
            <li>
              <NavLink to="categories" style={({isActive}) => (isActive ? activeStyle : undefined)}>
                <IoCopyOutline className="icon" />
                <span className="text">Categories</span>
              </NavLink>
            </li>
            <li>
              <NavLink to="/" style={({isActive}) => (isActive ? activeStyle : undefined)}>
                <IoIosMenu className="icon" />
                <span className="text">Home</span>
              </NavLink>
            </li>
            <li>
              <NavLink to="favorites" style={({isActive}) => (isActive ? activeStyle : undefined)}>
                <IoIosHeartEmpty className="icon" />
                <span className="text">Favorites</span>
              </NavLink>
            </li>
            <li>
              <NavLink to="profile" style={({isActive}) => (isActive ? activeStyle : undefined)}>
                <IoPersonOutline className="icon" />
                <span className="text">Profile</span>
              </NavLink>
            </li>
            {/* <div className="indicator"></div> */}
          </ul>
        </section>
      </div>
    </>
  );
}
```

Wie kann ich nun die Klassen einbauen? Der Code dazu liegt in der Navigation.css. [Hier mein aktuelles Repo dazu.](https://github.com/anresa/capstone-project/tree/navigation)

Vielen Dank schonmal!

**Reply by anresa** [reply]
@ThomasF85 @pixelass Ich glaube ich habe den richtigen Weg geraden gefunden. Meine (zweite) Frage kann erstmal hinten angestellt werden. Wenn ich es komplett schaffe, poste ich mein Lösung dazu später hier. Vielen Dank für eure Hilfe!

**Comment by pixelass** [comment]
Schau mal auf folgendes Kommentar: https://github.com/anresa/capstone-project/pull/11/files#r935859231

(Praktisch das gleiche was @ThomasF85 auch schon gesagt hat.)


Um `className` besser dynamisch zu kontrollieren kannst du mal schauen ob eine der folgenden libraries dir da hilft:

- https://github.com/JedWatson/classnames
- https://github.com/lukeed/clsx


Example:

```jsx
import clsx from 'clsx';

function MyComponent() {
  const [isActive, setIsActive] = useState(false);

  // className if isActive === false: "button"
  // className if isActive === true: "button active"
  return (
    <button
      className={clsx("button", {"active": isActive})}
      onClick={() => {
        setIsActive(!isActive)
      }}
    >
      Toggle this button
    </button>
  );
}
```
